File: sustainml_modules/test/communication/SimpleOrchestratorNode.py
# Copyright 2024 Proyectos y Sistemas de Mantenimiento SL (eProsima).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""SustainML Client Node API specification."""

import threading

from sustainml_swig import OrchestratorNodeHandle as cpp_OrchestratorNodeHandle
from sustainml_swig import OrchestratorNode as cpp_OrchestratorNode

from sustainml_swig import NodeStatus

class OrchestratorNodeHandle(cpp_OrchestratorNodeHandle):

    def __init__(self):
        self.node_data_received_ = {}
        # Condition variable
        self.cv_ = threading.Condition()

        # Parent class constructor
        super().__init__()

    # Callback
    def on_node_status_change(
            self,
            id: int,
            status : NodeStatus):

        if id not in self.node_data_received_:
            self.node_data_received_[id] = (0, 0)

        self.node_data_received_[id] = (status.node_status(), self.node_data_received_[id][1])

        self.cv_.acquire()
        self.cv_.notify()
        self.cv_.release()

    # Callback
    def on_new_node_output(
            self,
            id : int,
            data):

        if id not in self.node_data_received_:
            self.node_data_received_[id] = (0, 0)

        self.node_data_received_[id] = (self.node_data_received_[id][0], 1 + self.node_data_received_[id][1])

        self.cv_.acquire()
        self.cv_.notify()
        self.cv_.release()

        print("NodeOutput received ", id)

    def prepare_expected_data(
            self,
            expected_data):

        self.expected_data_ = expected_data

    def check_data(
            self):
        for key, value in self.node_data_received_.items():
            print(f"Arrived {key} {value[0]} {value[1]}")

        for key, value in self.expected_data_.items():
            print(f"Expected {key} {value[0]} {value[1]}")

        return self.expected_data_ == self.node_data_received_

    def wait_for_data(
            self):
        print("Waiting for data")
        with self.cv_:
            while not self.check_data():
                print("Waiting for data...")
                self.cv_.wait()
            print("Data received")

# Proxy class to instantiate by the user
class OrchestratorNode:

    def __init__(self):

        self.handler_ = OrchestratorNodeHandle()
        self.node_ = cpp_OrchestratorNode(self.handler_)

    # Proxy method to run the node
    def spin(self):

        self.node_.spin()

    # Proxy method to manually terminate
    def terminate():

        cpp_OrchestratorNode.terminate()

# Call main in program execution
if __name__ == '__main__':

    expected_data_nodes_ready = {
        # Status::NODE_IDLE (2), count (0)
        0: (2, 0), 1: (2, 0), 2: (2, 0), 3: (2, 0), 4: (2, 0), 5: (2, 0)
    }

    expected_data_task_executed = {
        # Status::NODE_IDLE (2), count (1)
        0: (2, 1), 1: (2, 1), 2: (2, 1), 3: (2, 1), 4: (2, 1), 5: (2, 1)
    }
    node = OrchestratorNode()
    node.handler_.prepare_expected_data(expected_data_nodes_ready)
    node.handler_.wait_for_data()
    node.handler_.prepare_expected_data(expected_data_task_executed)
    new_task = node.node_.prepare_new_task()
    task_id = new_task[0]
    ui = new_task[1]
    ui.task_id(task_id)
    ui.modality("video")
    ui.problem_definition("Classify cars in a video sequence.")
    node.node_.start_task(task_id, ui)
    node.handler_.wait_for_data()

File: sustainml_modules/test/communication/SimpleOrchestratorNode.cpp
// Copyright 2024 Proyectos y Sistemas de Mantenimiento SL (eProsima).
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <sustainml_cpp/orchestrator/OrchestratorNode.hpp>

struct SimpleOrchestratorNodeHandle : public sustainml::orchestrator::OrchestratorNodeHandle
{
    using DataCollection = std::map<sustainml::NodeID, std::pair<Status, int>>;

    void on_new_node_output(
            const sustainml::NodeID& id,
            void* data)
    {
        std::lock_guard<std::mutex> lock(mtx_);
        node_data_received_[id].second++;
        cv_.notify_one();
        std::cout << "NodeOutput received " << (int)id << std::endl;
    }

    void on_node_status_change(
            const sustainml::NodeID& id,
            const types::NodeStatus& status)
    {
        std::lock_guard<std::mutex> lock(mtx_);
        node_data_received_[id].first = status.node_status();
        cv_.notify_one();
    }

    void prepare_expected_data(
            const DataCollection& expected_data)
    {
        expected_node_data_ = expected_data;
    }

    template<class _Rep,
            class _Period
            >
    bool wait_for_data(
            const std::chrono::duration<_Rep, _Period>& max_wait)
    {
        std::unique_lock<std::mutex> lock(cv_mutex_);
        cv_.wait_for(lock, max_wait, [this]() -> bool
                {
                    for (auto& x : node_data_received_)
                    {
                        std::cout << "Arrived " << (int)x.first << (int)x.second.first << x.second.second << std::endl;
                    }
                    for (auto& x : expected_node_data_)
                    {
                        std::cout << "Expected " << (int)x.first << (int)x.second.first << x.second.second << std::endl;
                    }
                    return expected_node_data_ == node_data_received_;
                });

        return expected_node_data_ == node_data_received_;
    }

private:

    std::mutex cv_mutex_;
    std::condition_variable cv_;

    // map of NodeID to pair of <Status, Output counter>
    std::mutex mtx_;
    DataCollection node_data_received_;
    DataCollection expected_node_data_;
};

int main(
        int argc,
        char** argv)
{
    SimpleOrchestratorNodeHandle* sonh = new SimpleOrchestratorNodeHandle();

    // NodeID, <Node Status, Output counter>
    SimpleOrchestratorNodeHandle::DataCollection expected_data_nodes_ready =
    {
        {sustainml::NodeID::ID_ML_MODEL_METADATA, {(Status)2, 0}},
        {sustainml::NodeID::ID_ML_MODEL, {(Status)2, 0}},
        {sustainml::NodeID::ID_HW_RESOURCES, {(Status)2, 0}},
        {sustainml::NodeID::ID_CARBON_FOOTPRINT, {(Status)2, 0}},
        {sustainml::NodeID::ID_HW_CONSTRAINTS, {(Status)2, 0}},
        {sustainml::NodeID::ID_APP_REQUIREMENTS, {(Status)2, 0}}
    };

    // NodeID, <Node Status, Output counter>
    SimpleOrchestratorNodeHandle::DataCollection expected_data_task_executed =
    {
        {sustainml::NodeID::ID_ML_MODEL_METADATA, {(Status)2, 1}},
        {sustainml::NodeID::ID_ML_MODEL, {(Status)2, 1}},
        {sustainml::NodeID::ID_HW_RESOURCES, {(Status)2, 1}},
        {sustainml::NodeID::ID_CARBON_FOOTPRINT, {(Status)2, 1}},
        {sustainml::NodeID::ID_HW_CONSTRAINTS, {(Status)2, 1}},
        {sustainml::NodeID::ID_APP_REQUIREMENTS, {(Status)2, 1}}
    };

    sonh->prepare_expected_data(expected_data_nodes_ready);

    sustainml::orchestrator::OrchestratorNode orchestrator(*sonh);

    // First, wait for all nodes to be ready
    sonh->wait_for_data(std::chrono::hours(24));

    sonh->prepare_expected_data(expected_data_task_executed);

    auto new_task = orchestrator.prepare_new_task();
    new_task.second->task_id(new_task.first);
    new_task.second->modality("video");
    new_task.second->problem_definition("Classify cars in a video sequence.");

    orchestrator.start_task(new_task.first, new_task.second);

    // Now all nodes should have publish one output i.e the one for {1,1}
    sonh->wait_for_data(std::chrono::hours(24));

    return 0;
}
